apps/studio/src/lib/editor.ts

```diff
@@ -0,0 +1,131 @@
+import type { TableOrView } from "@/lib/db/models";
+import { DialectData } from "@shared/lib/dialects/models";
+
+interface Word {
+  text: string;
+  name: string;
+  type: "schema" | "table" | "column";
+  schema?: string;
+}
+
+/**
+ * A list of all words in key-value form, where the key is the word that's case sensitive.
+ */
+type WordList = Record<string, Word>;
+
+export interface DBHint {
+  defaultSchema?: string;
+  /**
+   * A list of tables in key-value form for fast queries, where the key
+   * is the table name. If the database supports schemas, all tables are from
+   * the default schema.
+   **/
+  defaultTableWordList: WordList;
+  /** list of all tables in array form. */
+  tableWords: Word[];
+  schemaWordList: WordList;
+  dialect: DialectData;
+}
+
+export function makeDBHint(
+  tables: TableOrView[],
+  dialect: DialectData,
+  defaultSchema?: string | null
+): DBHint {
+  const schemaSet = new Set<string>();
+  if (defaultSchema) schemaSet.add(defaultSchema);
+  if (defaultSchema === null) defaultSchema = undefined;
+
+  const tableWords: Word[] = [];
+
+  const defaultTableWordList = tables.reduce((acc, table) => {
+    if (table.schema) schemaSet.add(table.schema);
+
+    const word = {
+      name: table.name,
+      text: dialect.friendlyNormalizedIdentifier(table.name),
+      type: "table" as const,
+      schema: table.schema || defaultSchema,
+    };
+
+    tableWords.push(word);
+
+    const key = table.name;
+    if (table.schema && table.schema !== defaultSchema) return acc;
+
+    return { ...acc, [key]: word };
+  }, {});
+
+  const schemaWordList = Array.from(schemaSet).reduce(
+    (acc, name) => ({
+      ...acc,
+      [name]: {
+        name,
+        text: name,
+        type: "schema",
+      },
+    }),
+    {}
+  );
+
+  return {
+    tableWords,
+    schemaWordList,
+    defaultTableWordList,
+    defaultSchema,
+    dialect,
+  };
+}
+
+const quot = "'\"`";
+
+const SCHEMA_TABLE = new RegExp(
+  "(?:(?<schema>[quot]?.+?[quot]?)\\.(?<table1>[quot]?.+[quot]?)|(?<table2>[quot]?.+[quot]?))".replaceAll(
+    "quot",
+    quot
+  )
+);
```

Comment:
Unless you have a reason I'm not seeing here, could you replace `quot` with `quote`?